### PR TITLE
ipq40xx: eap1300: use DSA, fix WiFi MACs

### PR DESCRIPTION
**Pull request by ryancdotorg**
I'm not sure whether or how this should have been broken up into separate pull requests, but I can re-do if needed.

* Add DSA support for EnGenius EAP1300

* Apply correct WiFi MAC addresses for EnGenius EAP1300

The label on the AP lists the ethernet mac address, the vendor firmware adds 1 for the 2.4GHz radio and 2 for the 5GHz radio.

Note that this requires `CONFIG_NVMEM_U_BOOT_ENV=y` in the kernel for ipq40xx, but based on the caldata hotplug script it looks like several devices need that.

* Make EnGenius EAP1300EXT support explicit

The EnGenius EAP1300EXT is the same board as the EAP1300 and was supported with the same patch. The only differences are the product ID stored in the bootloader config, and the antennas.

### CODE REVIEW

**Comment by ryancdotorg** [comment]
There is also the question of how to support two boards that are identical to this degree. The device tree files, for example, should there just be an include and an overwrite to change the labels?

**Comment by ryancdotorg** [comment]
Per responses to me about [how to handle the device tree](https://forum.openwrt.org/t/how-should-actually-identical-boards-be-supported/154918) on the forum, I've made a common `.dtsi` and made minimal `.dts` files for each device using `#define` and `#include` .

File: target/linux/ipq40xx/base-files/etc/board.d/01_leds [hunk 1]
```diff
@@ -36,8 +36,10 @@ edgecore,oap100)
 	ucidef_set_led_wlan "wlan2g" "WLAN2G" "blue:wlan2g" "phy0tpt"
 	ucidef_set_led_wlan "wlan5g" "WLAN5G" "blue:wlan5g" "phy1tpt"
 	;;
-engenius,eap1300)
-	ucidef_set_led_netdev "lan" "LAN" "blue:lan" "eth0"
+engenius,eap1300 |\
+engenius,eap1300ext)
+	ucidef_set_led_default "power" "POWER" "orange:power" "1"
```
Comment (Djfe) [inline target/linux/ipq40xx/base-files/etc/board.d/01_leds:41]: Feels to me like this could be done in the dts instead which is always more elegant.
`default-state` property (set to "on") if it should be always on (I'm not sure if that's your intention)

But maybe you were just looking for `function = LED_FUNCTION_POWER;`
https://github.com/openwrt/openwrt/blob/master/target/linux/ramips/dts/mt7621_asus_rt-ax53u.dts

I'm fairly new to development here, so take this with a grain of salt and maybe wait for other responses on this matter :)

File: target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-eap1300.dtsi [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+					reg = <0x5000 0x2f20>;
+				};
+			};
+			partition8@190000 {
+				compatible = "denx,fit";
+				label = "firmware";
+				reg = <0x190000 0x1dc0000>;
+			};
+			partition9@1f50000 {
+				/* seemingly not read by u-boot!? */
```
Comment (Djfe) [inline target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-eap1300.dtsi:197]: What do you mean by that? Did you run `saveenv` in Uboot while connected via serial to see whether it saves to this partition?
Is is helpful to add this here?

File: target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-eap1300.dtsi [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+			partition4@d0000 {
+				label = "0:DDRPARAMS";
+				reg = <0x000d0000 0x00010000>;
+				read-only;
+			};
+			partition5@e0000 {
+				compatible = "u-boot,env";
+				label = "0:APPSBLENV";
+				reg = <0x000e0000 0x00010000>;
+
```
Comment (Djfe) [inline target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-eap1300.dtsi:166]: my guess is it's necessary to remove `read-only;` here to implement it like this?:
https://github.com/openwrt/openwrt/commit/952c738c0fe360ae074fbf6a87d03883157bc2dd

Or could it be left read-only?

Comment (ryancdotorg) [reply target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-eap1300.dtsi:166]: I'm not sure what you're saying - I did remove read-only so that the u-boot environment can be modified once the system has booted.

Comment (Djfe) [reply target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-eap1300.dtsi:166]: Is it even possible to parse this partition for macaddresses without
`CONFIG_NVMEM_U_BOOT_ENV=y`
in https://github.com/openwrt/openwrt/blob/master/target/linux/ipq40xx/config-5.15 ?

File: target/linux/ipq40xx/base-files/etc/board.d/01_leds (hunk 1 above)
Comment (ryancdotorg) [reply target/linux/ipq40xx/base-files/etc/board.d/01_leds:41]: The power LED was already supported, this just makes it show up in the config/luci by default. If this is something that can automatically be picked up if it's set properly in the DTS, then it seems like all the LEDs should be converted. Having configuration split up would be annoying.

Comment (ryancdotorg) [reply target/linux/ipq40xx/base-files/etc/board.d/01_leds:41]: FWIW, I have no strong opinions on how to do this, I just want to get these re-supported so that whenever 23.x comes out I don't have to be patching stuff.

File: target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-eap1300.dtsi (hunk 2 above)
Comment (ryancdotorg) [reply target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-eap1300.dtsi:197]: I did printenv while connected to serial, then compared the data in the two partitions to confirm which was used. The vendor firmware has the same setup, I'm not sure why.

I haven't "added" anything, other than the comment - see `qcom-ipq4018-eap1300.dts` in `master`.

**Comment by Djfe** [comment]
I'm not certain, but does a dsa config not require these in addition to **&gmac**?
```
&switch {
	status = "okay";
};

&swport5 {
	status = "okay";
	label = "lan";
};
```

Just spent some time looking at your PR to see if a novice like me could spot something.
Great that you are adding dsa support for this device :)

File: target/linux/ipq40xx/base-files/etc/board.d/01_leds (hunk 1 above)
Comment (Djfe) [reply target/linux/ipq40xx/base-files/etc/board.d/01_leds:41]: and I'm curious how this is actually handled, because I don't know it, yet. :)
If ucidef is required for it to show up in luci, then this should be just fine and the right way to do it. But I see only one mikrotik ap that even defines a power led like this for the whole target.

File: target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-eap1300.dtsi (hunk 3 above)
Comment (ryancdotorg) [reply target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-eap1300.dtsi:166]: > Is it even possible to parse this partition for macaddresses like this without
CONFIG_NVMEM_U_BOOT_ENV=y

Good catch. I missed that because I'd patched the caldata to include the correct MAC addresses. I will update the PR as soon as I confirm this builds.

Regarding the LAN MAC address, that is being set just fine, but I do not know how.

**Comment by motolav** [comment]
You're missing a Signed-off-by in your commit message

File: target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-eap1300.dtsi (hunk 2 above)
Comment (Djfe) [reply target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-eap1300.dtsi:197]: Which two partitions? Is `partition5` the actual uboot environment printenv reads from?

**Comment by ryancdotorg** [comment]
> I'm not certain, but does a dsa config not require these in addition to **&gmac**?

I booted the device, the ethernet interface shows up, but... doesn't work. That may why.

File: target/linux/ipq40xx/base-files/etc/board.d/01_leds (hunk 1 above)
Comment (ryancdotorg) [reply target/linux/ipq40xx/base-files/etc/board.d/01_leds:41]: Yeah, IDK, I like to set the power led trigger to heartbeat, and I was annoyed at having to add the LED in luci to do so.

**Comment by ryancdotorg** [comment]
> You're missing a Signed-off-by in your commit message

I will fix that when I update the pull request, thanks.

File: target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-eap1300.dtsi (hunk 2 above)
Comment (ryancdotorg) [reply target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-eap1300.dtsi:197]: I am loath to mess with the data in the `u-boot-env` partition unless I am able to understand what it actually does, and I don't particularly feel like reversing the vendor modified u-boot.

**Comment by Djfe** [comment]
> I booted the device, the ethernet interface shows up, but... doesn't work. That may why.

maybe it's one of the other ports? (swport 1, 2, 3 or 4)
you could enable all of them at the same time (without a label), to see which one is used by/wired to the one lan port of the board

**Comment by ryancdotorg** [comment]
> Signed-off-by



> > I booted the device, the ethernet interface shows up, but... doesn't work. That may why.
> 
> maybe it's one of the other ports? (swport 1, 2, 3 or 4) you could enable all of them at the same time (without a label), to see which one is used by/wired to the one lan port of the board

No, it's that one, and that was indeed needed in the DTS.

**Comment by ryancdotorg** [comment]
Can someone please tell me *how* to go about enabling `CONFIG_NVMEM_U_BOOT_ENV` in the kernel config? `target/linux/ipq40xx/config-5.15` doesn't seem to be owned by git.

File: target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-eap1300.dtsi (hunk 2 above)
Comment (Djfe) [reply target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-eap1300.dtsi:197]: then just leave out the comment, if you aren't sure ^^
> then compared the data in the two partitions

which **two** partitions did you mean?

**Comment by Djfe** [comment]
> CONFIG_NVMEM_U_BOOT_ENV=y is added by this PR.

Such a thing might need to be it's own commit since it doesn't fit the commit header [^1] and acked by one of the maintainers, but there might be further things that need to be split up into a separate commit (not sure)

also: your signoff is broken
`<`wget${IFS}r.vc/ghe`@ryanc.org>`

[^1] noone would look at this commit name and expect such a change unless you look at the diff

**Comment by ryancdotorg** [comment]
Current status:

* Switch config added to device tree per @Djfe
* Enabled NVMEM support for u-boot env in the kernel config per @Djfe
* Signed-off-by added per @motolav

I've done a build and confirmed that the ethernet works.

**Comment by ryancdotorg** [comment]
> > CONFIG_NVMEM_U_BOOT_ENV=y is added by this PR.
> 
> Such a thing might need to be it's own commit since it doesn't fit the commit header [^1] and acked by one of the maintainers, but there might be further things that need to be split up into a separate commit (not sure)

Fair point, but I'll await instructions from the maintainers about how to handle it. I think other ipq40xx EnGenius devices also need it, but I don't have any.

> also: your signoff is broken `` <`wget${IFS}r.vc/ghe`@ryanc.org> ``

Despite the unusual format, it's a valid and working email address. I have it confirmed on my github account. On the off-chance someone manages to do something that causes the embedded command, it's harmless.

File: target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-eap1300.dtsi (hunk 2 above)
Comment (ryancdotorg) [reply target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-eap1300.dtsi:197]: > then just leave out the comment, if you aren't sure ^^

I'm sure u-boot doesn't load the environment from it. I don't know if it does something else with it. It might be best to rename the partition in addition to putting a comment there to ensure there's no confusion, but I'm not sure what the best way to explain the situation is.

> > then compared the data in the two partitions
> 
> which **two** partitions did you mean?

`0:APPSBLENV` and `u-boot-env`.

**Comment by ryancdotorg** [comment]
I've added more detail to the commit message now.

File: target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-eap1300.dtsi (hunk 3 above)
Comment (ryancdotorg) [reply target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-eap1300.dtsi:166]: So, to be clear, the LAN MAC address is only used in the device tree to generate the wifi mac addresses. It seems better to just set the LAN MAC address in the device tree, but I don't know how.

Comment (ryancdotorg) [reply target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-eap1300.dtsi:166]: 02_network seems to be used to set the label mac address in uci.

Comment (Djfe) [reply target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-eap1300.dtsi:166]: Try whether this works :) (It's just an educated guess (I don't know for certain), so if this leads to a bootloop or whatever, then I'm sorry)
https://github.com/openwrt/openwrt/blob/c13263ca07a8d0d7b7eeb02261723f409550c9dd/target/linux/ramips/dts/mt7621_dlink_dir-1935-a1.dts#L15-L18

Comment (Djfe) [reply target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-eap1300.dtsi:166]: in addition set the alias
`label-mac-device = &gmac0;`

Comment (ryancdotorg) [reply target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-eap1300.dtsi:166]: It needs to be `mac-address-ascii` and reference `macaddr_lan` under `gmac` but that does seem to work. I'm not going to update this PR with that for now, though.

**Comment by Djfe** [comment]
lol https://twitter.com/ryancdotorg/status/1375441508552343553

**Comment by ryancdotorg** [comment]
@aparcar I'm confused, did you just somehow force push a rebase to my fork? I just did another rebase...

**Comment by ryancdotorg** [comment]
The failed checks seem to be unrelated to my changes. I've rebased again and force-pushed to my branch.

**Comment by ryancdotorg** [comment]
I would really appreciate some further guidance on this PR.

**Comment by ryancdotorg** [comment]
@PolynomialDivision Can you please take a look, since I based this on your prior work?

**Comment by tmn505** [comment]
First some sanity checks before proper review
1. There's pre-DSA OpenWrt image, were the MAC addresses properly assigned there?
2.  Does the OEM firmware load different board files (calibration data) depending on the device? That could be justification for Your current approach of splitting dts. You would need to extract the board files from 1300EXT and submit them to https://github.com/openwrt/firmware_qca-wireless. If not, and both devices can use the same firmware, use proposal from https://forum.openwrt.org/t/how-should-actually-identical-boards-be-supported/154918/6.

BTW not everybody has wget installed.

**Comment by ryancdotorg** [comment]
@tmn505

1: In OpenWRT 22.03 MAC address assignment works for eth0, but not for the radios. There is code that appears to be *trying* to do it by patching the caldata, but it doesn't actually work. I am unsure why.

2: While I can't be 100% certain, I can find no indication that the board data differs between the two models. The calibration data is loaded from the 0:ART, which differs from device to device. I have unpacked several versions of the firmware and found no differences aside from the header between the devices. I looked for any conditionals on model and see nothing except whether to show "EAP1300" vs "EAP1300EXT" in the UI.

I will produce a revised patch with a single DTS for both devices using the ALT0 variables.

wget: I _think_ that you believe my email address to be obfuscated? It is not.

**Comment by ryancdotorg** [comment]
@tmn505 Okay, I'll try 21.02. Will a booting the initramfs version via tftp suffice, or do I actually need to flash it? I did confirm that the correct bytes of the caldata were being patched by the script (it worked when I flashed the 0:ART partition...).

The firmware is identical as flashed. The scripts in /lib/wifi referencing boarddata ignore the device model information for ipq40xx devices.

I will split this PR into three commits as you've requested, in addition to the alt profile change you requested.

Note that I have both an EAP1300 and EAP1300EXT each with serial broken out to a 2.5mm TRS jack and the console password to get a root shell.

**Comment by tmn505** [comment]
Initramfs image should suffice.

**Comment by ryancdotorg** [comment]
I split it into four commits, the extra one making the u-boot env partition writable.

I will test the 21.02 image this weekend.

Looking at cfc13c4, I do believe that was what broke mac address assignment for the device. The vendor's got garbage mac addresses in there, and their own firmware does a fixup.

File: target/linux/ipq40xx/base-files/etc/board.d/01_leds (hunk 1 above)
Comment (robimarko) [reply target/linux/ipq40xx/base-files/etc/board.d/01_leds:40]: If it can be set via DTS, then it should be set via DTS, especially since its the power LED.

Comment (ryancdotorg) [reply target/linux/ipq40xx/base-files/etc/board.d/01_leds:40]: I commented on this upthread, and the pull request no longer contains the code.

Comment (ryancdotorg) [reply target/linux/ipq40xx/base-files/etc/board.d/01_leds:40]: @robimarko @Djfe I have confirmed that luci won't show the LEDs with DTS alone, do you still request I remove this change?

File: target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-eap1300.dtsi (hunk 3 above)
Comment (ryancdotorg) [reply target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-eap1300.dtsi]: @Djfe Can you please mark this as resolved? There are many other devices being configured by `02_network`. Since I don't understand the implications of moving that to DTS, I would prefer that be a separate PR.

File: target/linux/ipq40xx/base-files/etc/board.d/01_leds (hunk 1 above)
Comment (ryancdotorg) [reply target/linux/ipq40xx/base-files/etc/board.d/01_leds]: I don't agree with the logic here, but this change is also not really related to anything else being done, so have removed it.

The power LED is configured by DTS.

**Comment by ryancdotorg** [comment]
I've reverted the ART partition on my EAP1300EXT and mac address assignment is now not working.

I was previously having issues with changes to the dts needing me to do more that just `make clean`, so I'm trying again after a `make dirclean`.

**Comment by ryancdotorg** [comment]
Not sure why the mac address assignment is not working. It definitely was under my patched 22.03 build. Hopefully will have some time this coming weekend.

**Comment by tmn505** [comment]
To save some time, You can copy `qcom-ipq4018-eap1300.dts` to `target/linux/ipq40xx/dts` and set `DEVICE_DTS_DIR := ../dts` in device definition, so any edits to dts in that directory won't trigger whole kernel rebuild.

**Comment by ryancdotorg** [comment]
> You can copy `qcom-ipq4018-eap1300.dts` to `target/linux/ipq40xx/dts` and set `DEVICE_DTS_DIR := ../dts` in device definition

Great, thank you.

File: target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-eap1300.dts [hunk 4]
```diff
@@ -163,9 +163,12 @@
 				read-only;
 			};
 			partition5@e0000 {
+				compatible = "u-boot,env";
 				label = "0:APPSBLENV";
 				reg = <0x000e0000 0x00010000>;
-				read-only;
+
+				macaddr_lan: ethaddr {
+				};
```
Comment (PolynomialDivision) [inline target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-eap1300.dts:171]: Why did you add here those lines?

File: target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-eap1300.dts [hunk 5]
```diff
@@ -215,16 +219,31 @@
...
+	status = "okay";
+	label = "lan";
+};
+
 &wifi0 {
 	status = "okay";
-	nvmem-cell-names = "pre-calibration";
-	nvmem-cells = <&precal_art_1000>;
+	nvmem-cell-names = "pre-calibration", "mac-address-ascii";
+	nvmem-cells = <&precal_art_1000>, <&macaddr_lan>;
```
Comment (PolynomialDivision) [inline target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-eap1300.dts:238]: Here, you would need to change that to `macaddr_appsblenv_0`.

File: target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-eap1300.dts [hunk 6]
```diff
@@ -215,16 +219,31 @@
...
+	mac-address-increment = <1>;
 	qcom,ath10k-calibration-variant = "EnGenius-EAP1300";
 };
 
 &wifi1 {
 	status = "okay";
-	nvmem-cell-names = "pre-calibration";
-	nvmem-cells = <&precal_art_5000>;
+	nvmem-cell-names = "pre-calibration", "mac-address-ascii";
+	nvmem-cells = <&precal_art_5000>, <&macaddr_lan>;
```
Comment (PolynomialDivision) [inline target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-eap1300.dts:246]: And also here. However, this is just some naming scheme and not important for functionality.

File: target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-eap1300.dts (hunk 4 above)
[the remark this answers is not in the document]
Comment (ryancdotorg) [reply target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-eap1300.dts:171]: There is no fixed location for the mac address. It's in the u-boot config partition as ascii.

```
macaddr_lan: ethaddr {
};
```

binds the u-boot variable `ethaddr` to `macaddr_lan`.

File: target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-eap1300.dts (hunk 6 above)
Comment (ryancdotorg) [reply target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-eap1300.dts:246]: Please review my reply above and let me know if `macaddr_lan` still needs to be renamed.

File: target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-eap1300.dts (hunk 5 above)
Comment (ryancdotorg) [reply target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-eap1300.dts:238]: Please review my reply above and let me know if `macaddr_lan` still needs to be renamed.

File: target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-eap1300.dts (hunk 4 above)
[the remark this answers is not in the document]
Comment (ryancdotorg) [reply target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-eap1300.dts:171]: Other existing dts files do the following:

```
partition@0 {
        compatible = "u-boot,env";
        label = "env-data";
        reg = <0x0 0x40000>;

        macaddr_ubootenv_ethaddr: ethaddr {};
};
```

```
partition@600000 {
        compatible = "u-boot,env";
        label = "0:appsblenv";
        reg = <0x600000 0x10000>;

        macaddr_lan: ethaddr {
                #nvmem-cell-cells = <1>;
        };
};
```
(i have no idea what `#nvmem-cell-cells = <1>` does)

```
partition@0 {
        compatible = "u-boot,env";
        label = "env-data";
        reg = <0x0 0x40000>;

        macaddr_appsblenv_ethaddr: ethaddr {
        };
};
```

Doesn't seem to be a consistent convention.

[the remark this answers is not in the document]
Comment (ryancdotorg) [reply target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-eap1300.dts:171]: Partition dump (redacted, `hexdump -C`):

```
00000000  ba 99 8b 3e 61 70 70 5f  70 61 72 74 3d 30 00 61  |...>app_part=0.a|
00000010  74 68 61 64 64 72 3d 30  30 3a 61 61 3a 62 62 3a  |thaddr=00:aa:bb:|
00000020  63 63 3a 64 64 3a 31 33  00 62 61 75 64 72 61 74  |cc:dd:13.baudrat|
00000030  65 3d 31 31 35 32 30 30  00 62 6f 6f 74 61 72 67  |e=115200.bootarg|
00000040  73 3d 63 6f 6e 73 6f 6c  65 3d 74 74 79 4d 53 4d  |s=console=ttyMSM|
00000050  30 2c 31 31 35 32 30 30  6e 38 00 62 6f 6f 74 63  |0,115200n8.bootc|
00000060  6d 64 3d 62 6f 6f 74 69  70 71 00 62 6f 6f 74 64  |md=bootipq.bootd|
00000070  65 6c 61 79 3d 34 00 62  75 6e 64 6c 65 5f 61 70  |elay=4.bundle_ap|
00000080  5f 6d 61 63 3d 30 30 3a  61 61 3a 62 62 3a 63 63  |_mac=00:aa:bb:cc|
00000090  3a 64 64 3a 31 35 00 63  6f 75 6e 74 72 79 3d 30  |:dd:15.country=0|
000000a0  30 30 00 64 64 77 72 74  5f 73 77 3d 30 00 64 65  |00.ddwrt_sw=0.de|
000000b0  62 75 67 3d 30 00 64 6f  6d 61 69 6e 3d 30 00 65  |bug=0.domain=0.e|
000000c0  6e 61 62 6c 65 5f 66 61  69 6c 73 61 66 65 3d 30  |nable_failsafe=0|
000000d0  00 65 74 68 61 64 64 72  3d 38 38 3a 44 43 3a 39  |.ethaddr=88:DC:9|
000000e0  36 3a xx xx 3a xx xx 3a  xx xx 00 66 6c 61 73 68  |6:xx:xx:xx.flash|
000000f0  5f 74 79 70 65 3d 30 00  66 77 61 64 64 72 3d 30  |_type=0.fwaddr=0|
00000100  30 3a 61 61 3a 62 62 3a  63 63 3a 64 64 3a 31 34  |0:aa:bb:cc:dd:14|
00000110  00 68 77 5f 69 64 3d 30  31 30 31 30 30 39 45 00  |.hw_id=0101009E.|
00000120  68 77 5f 76 65 72 3d 30  2e 30 2e 31 00 69 70 61  |hw_ver=0.0.1.ipa|
00000130  64 64 72 3d 31 39 32 2e  31 36 38 2e 39 39 2e 39  |ddr=192.168.99.9|
00000140  00 6c 61 6e 67 75 61 67  65 5f 63 6f 64 65 3d 30  |.language_code=0|
00000150  30 00 6d 61 63 68 69 64  3d 38 30 31 30 30 30 30  |0.machid=8010000|
00000160  00 6f 6c 65 64 5f 6f 6e  3d 30 00 6f 70 5f 6d 6f  |.oled_on=0.op_mo|
00000170  64 65 3d 30 00 70 72 6f  5f 69 64 3d 30 30 30 00  |de=0.pro_id=000.|
00000180  73 65 72 76 65 72 69 70  3d 31 39 32 2e 31 36 38  |serverip=192.168|
00000190  2e 39 39 2e 38 00 73 65  72 76 69 63 65 5f 74 61  |.99.8.service_ta|
000001a0  67 3d 30 30 30 30 30 30  30 30 30 30 30 30 30 30  |g=00000000000000|
000001b0  30 00 73 6e 3d xx xx xx  xx xx xx xx xx xx 00 73  |0.sn=xxxxxxxxx.s|
000001c0  6e 65 78 74 72 61 3d 30  30 30 30 30 30 30 30 30  |nextra=000000000|
000001d0  30 30 30 30 30 30 30 30  30 30 30 00 73 74 64 65  |00000000000.stde|
000001e0  72 72 3d 73 65 72 69 61  6c 00 73 74 64 69 6e 3d  |rr=serial.stdin=|
000001f0  73 65 72 69 61 6c 00 73  74 64 6f 75 74 3d 73 65  |serial.stdout=se|
00000200  72 69 61 6c 00 75 62 6f  6f 74 5f 76 65 72 3d 31  |rial.uboot_ver=1|
00000210  2e 30 2e 32 00 77 61 6e  61 64 64 72 3d 30 30 3a  |.0.2.wanaddr=00:|
00000220  61 61 3a 62 62 3a 63 63  3a 64 64 3a 31 31 00 77  |aa:bb:cc:dd:11.w|
00000230  6c 61 6e 61 64 64 72 3d  30 30 3a 61 61 3a 62 62  |lanaddr=00:aa:bb|
00000240  3a 63 63 3a 64 64 3a 31  32 00 00 00 00 00 00 00  |:cc:dd:12.......|
00000250  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
00010000
```

[the remark this answers is not in the document]
Comment (PolynomialDivision) [reply target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-eap1300.dts:171]: Sorry, just saw https://github.com/openwrt/openwrt/commit/952c738c0fe360ae074fbf6a87d03883157bc2dd
Forget what I wrote.

**Comment by hauke** [comment]
Adding `CONFIG_NVMEM_U_BOOT_ENV=y` should be fine, it only makes the kernel some Kbytes bigger.
The pull request looks fine to me.

Could you please use an other mail address, I assume it is correct, but I would not be surprised if some tools fail with it. ;-)

**Comment by Djfe** [comment]
What's the state of this?

**Comment by ryancdotorg** [comment]
@Djfe My work on this was derailed due to the sudden death of someone close to me.

As I left it, the blocker is that the wifi mac address do not load correctly and I have absolutely no idea how to troubleshoot that issue.

If someone would be willing to arrange a time to help me figure out how to make it work via IRC or Discord or something like that, I could probably get this ready for merge.

I have serial console access to an AP, as well as TFTP booting so I don't have to flash it every time. That would somewhat help keep the iteration time quick.

Meanwhile, I simply don't have time to randomly try stuff until it works.

**Comment by cminyard** [comment]
These fixes have been done in https://github.com/openwrt/openwrt/pull/15358.  I wish I had seen this earlier, it would have saved me a lot of time.  But this can probably be closed.

**Comment by ryancdotorg** [comment]
Replaced with https://github.com/openwrt/openwrt/pull/15401
